javascript/ql/lib/semmle/javascript/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -1,10 +1,70 @@
 private import javascript
+private import semmle.javascript.dataflow.internal.CallGraphs
 private import semmle.javascript.dataflow.internal.DataFlowNode
+private import semmle.javascript.dataflow.internal.FlowSteps as FlowSteps
+private import semmle.javascript.dataflow.internal.Contents::Private
+private import semmle.javascript.dataflow.internal.VariableCapture
+private import semmle.javascript.dataflow.internal.sharedlib.DataFlowImplCommon as DataFlowImplCommon
 
-class Node = DataFlow::Node;
+private class Node = DataFlow::Node;
 
 class PostUpdateNode = DataFlow::PostUpdateNode;
 
+cached
+newtype TReturnKind =
+  MkNormalReturnKind() or
+  MkExceptionalReturnKind()
+
+class ReturnKind extends TReturnKind {
+  string toString() {
+    this = MkNormalReturnKind() and result = "return"
+    or
+    this = MkExceptionalReturnKind() and result = "exception"
+  }
+}
+
+private predicate returnNodeImpl(DataFlow::Node node, ReturnKind kind) {
+  node instanceof TFunctionReturnNode and kind = MkNormalReturnKind()
+  or
+  exists(Function fun |
+    node = TExceptionalFunctionReturnNode(fun) and
+    kind = MkExceptionalReturnKind() and
+    // For async/generators, the exception is caught and wrapped in the returned promise/iterator object.
+    // See the models for AsyncAwait and Generator.
+    not fun.isAsyncOrGenerator()
+  )
+}
+
+private DataFlow::Node getAnOutNodeImpl(DataFlowCall call, ReturnKind kind) {
+  kind = MkNormalReturnKind() and result = call.asOrdinaryCall()
+  or
+  kind = MkExceptionalReturnKind() and result = call.asOrdinaryCall().getExceptionalReturn()
+  or
+  kind = MkNormalReturnKind() and result = call.asBoundCall(_)
+  or
+  kind = MkExceptionalReturnKind() and result = call.asBoundCall(_).getExceptionalReturn()
+  or
+  kind = MkNormalReturnKind() and result = call.asAccessorCall().(DataFlow::PropRead)
+}
+
+class ReturnNode extends DataFlow::Node {
+  ReturnNode() { returnNodeImpl(this, _) }
+
+  ReturnKind getKind() { returnNodeImpl(this, result) }
+}
+
+/** A node that receives an output from a call. */
+class OutNode extends DataFlow::Node {
+  OutNode() { this = getAnOutNodeImpl(_, _) }
+}
+
+OutNode getAnOutNode(DataFlowCall call, ReturnKind kind) { result = getAnOutNodeImpl(call, kind) }
```

Comment:
It's annoying but no. `getAnOutNode` implements this part of `InputSig` to `DataFlowMake`:
```codeql
signature module InputSig {
  OutNode getAnOutNode(DataFlowCall call, ReturnKind kind);
}
```
We therefore must use the return type `OutNode` in this predicate. It would thus lead to empty recursion if we also used it to define the charpred of `OutNode`.